Mathlib/RingTheory/Polynomial/Hilbert.lean

```diff
@@ -0,0 +1,92 @@
+/-
+Copyright (c) 2024 Fangming Li. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fangming Li, Jujian Zhang
+-/
+import Mathlib.Algebra.Polynomial.AlgebraMap
+import Mathlib.Algebra.Polynomial.Div
+import Mathlib.Algebra.Polynomial.Eval.SMul
+import Mathlib.Data.Nat.Factorial.BigOperators
+import Mathlib.RingTheory.PowerSeries.WellKnown
+
+/-!
+# Hilbert polynomials
+
+In this file, we aim to formalise a useful fact: given any `p : ℤ[X]` and `d : ℕ`, there exists
+some `h : ℚ[X]` such that for any large enough `n : ℕ`, `h(n)` is equal to the coefficient of `Xⁿ`
+in the power series expansion of `p/(1 - X)ᵈ` (or `p * (PowerSeries.invOneSubPow ℤ d)`). This `h` is
+unique and is called the Hilbert polynomial with respect to `p` and `d` (`Polynomial.hilbert p d`).
+
+The above fact is used to construct the Hilbert polynomial of a graded module that satisfies certain
+conditions. Specifically:
+* Assume `A = ⊕ₙAₙ` is a Noetherian ring graded by `ℕ` such that `A` is generated by `a₁,...,aₛ` as
+  an `A₀`-algebra, where for each `i = 1,...,s`, `aᵢ` is a homogeneous element in `A` of degree
+  `dᵢ > 0`. Let `M = ⊕ₙMₙ` be a finitely generated `A`-module graded by `ℕ`. Then it is true that
+  each `Mₙ` is a finitely generated module over `A₀`.
+* Let `λ : C → ℤ` be an additive function, where `C` is the collection of all finitely generated
+  `A₀`-modules; in other words, given any short exact sequence `0 ⟶ N ⟶ O ⟶ P ⟶ 0` of finitely
+  generated modules over `A₀`, we have `λ(O) = λ(N) + λ(P)`. Then the Poincaré series of `M` in
+  terms of `λ` is the formal power series `P(M, λ, X) = Σₙλ(Mₙ)Xⁿ`. The Hilbert-Serre Theorem states
+  that `P(M, λ, X)` can be written as `p(X)/∏ᵢ₌₁,...,ₛ(1 - X ^ dᵢ)`, where `p(X)` is a polynomial
+  with coefficients in `ℤ`.
+* For the case when `d₁,...,dₛ = 1`, the Poincaré series of `M` with respect to `λ` can be expressed
+  as `p(X)/(1 - X)ˢ`. Hence the fact stated in the beginning guarantees an `h : ℚ[X]` such that for
+  any large enough `n : ℕ`, the coefficient of `Xⁿ` in `P(M, λ, X)` equals `h.eval (n : ℚ)`. This
+  `h` is called the Hilbert polynomial of `M` in terms of `λ`, which is what we eventually want to
+  formalise.
+-/
```

Comment:
Most of this module docstring is nothing to do with what is in the file. Also, the docstring does not follow the guidelines for module docstrings https://leanprover-community.github.io/contribute/doc.html at all. Here is a suggestion for a better module docstring. 

```
/-!
# Hilbert polynomials

In this file, we formalise the following statement: given any `p : ℤ[X]` and `d : ℕ`, there exists
some `h : ℚ[X]` such that for any large enough `n : ℕ`, `h(n)` is equal to the coefficient of `Xⁿ`
in the power series expansion of `p/(1 - X)ᵈ`. This `h` is unique and is called the
Hilbert polynomial of `p` and `d` (`Polynomial.hilbert p d`).

## Main definitions

* `Polynomial.hilbert p d`. If `p : ℤ[X]` and `d : ℕ` then `Polynomial.hilbert p d : ℚ[X]`
  is the polynomial whose value at `n` equals the coefficient of `Xⁿ` in the power
  series expansion of `p/(1 - X)ᵈ`

## TODO

* Prove that `Polynomial.hilbert p d : ℚ[X]` is the polynomial whose value at `n` equals the
coefficient of `Xⁿ` in the power series expansion of `p/(1 - X)ᵈ`

* Hilbert polynomials of graded modules.
-/
```

Also, it is not clear to me why you restrict to `p : ℤ[X]` when clearly `p : ℚ[X]` would work just as well. In fact why not do `p : k[X]` for `k` any field?

Reply:
> Most of this module docstring is nothing to do with what is in the file. Also, the docstring does not follow the guidelines for module docstrings https://leanprover-community.github.io/contribute/doc.html at all. Here is a suggestion for a better module docstring.
> 
> ```
> /-!
> # Hilbert polynomials
> 
> In this file, we formalise the following statement: given any `p : ℤ[X]` and `d : ℕ`, there exists
> some `h : ℚ[X]` such that for any large enough `n : ℕ`, `h(n)` is equal to the coefficient of `Xⁿ`
> in the power series expansion of `p/(1 - X)ᵈ`. This `h` is unique and is called the
> Hilbert polynomial of `p` and `d` (`Polynomial.hilbert p d`).
> 
> ## Main definitions
> 
> * `Polynomial.hilbert p d`. If `p : ℤ[X]` and `d : ℕ` then `Polynomial.hilbert p d : ℚ[X]`
>   is the polynomial whose value at `n` equals the coefficient of `Xⁿ` in the power
>   series expansion of `p/(1 - X)ᵈ`
> 
> ## TODO
> 
> * Prove that `Polynomial.hilbert p d : ℚ[X]` is the polynomial whose value at `n` equals the
> coefficient of `Xⁿ` in the power series expansion of `p/(1 - X)ᵈ`
> 
> * Hilbert polynomials of graded modules.
> -/
> ```
> 
> Also, it is not clear to me why you restrict to `p : ℤ[X]` when clearly `p : ℚ[X]` would work just as well. In fact why not do `p : k[X]` for `k` any field?

I have figured out a way to generalise my work to any field `F` with `CharZero F`, but I'm not sure if the generalised version of my codes will still fit well with @jjaassoonn's work in #9819? In his statement of the Hilbert-Serre Theorem, the polynomial `p` in the expression `p/(1 - X)ᵈ` is assumed to be in `ℤ[X]`. Then I guess I need to find some way to let Lean know that `ℤ[X]` is a subset of `ℚ[X]`.

Reply:
> You can just `Polynomial.map` it into `ℚ[X]`

OK I have pushed the generalised version of my codes.